services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -201,7 +203,28 @@ public Response processGrantRequest() {
 
         switch (action) {
             case AUTHORIZATION_CODE:
-                return codeToToken();
+                // Code to token needs to be run in a retriable transaction as concurrent execution of this action can lead
+                // to exceptions on DBs with SERIALIZABLE isolation level. See ConcurrentLoginTest#concurrentCodeReuseShouldFail.
+                // If other actions lead to the same issue we might broaden the scope of the retriable block and do it for all actions.
+                Object result = KeycloakModelUtils.runJobInRetriableTransaction(this.session.getKeycloakSessionFactory(), kcSession -> {
+                    try {
+                        RealmModel realmModel = kcSession.realms().getRealm(realm.getId());
+                        ClientModel clientModel = realmModel.getClientByClientId(client.getClientId());
+                        kcSession.getContext().setRealm(realmModel);
+                        kcSession.getContext().setClient(clientModel);
+                        return codeToToken(kcSession, realmModel, clientModel);
+                    } catch (WebApplicationException we) {
+                        // WebApplicationException needs to be returned and treated (rethrown) by the calling code because the new transaction
+                        // still needs to be committed when this exception is thrown. It captures final business states that won't change when
+                        // being retried, like an invalid code.
+                        return we;
+                    }
```

Comment:
Given the separation proposed in this PR, could the whole code-to-token logic be moved to a private _static_ class within `TokenEndpoint`?

As of now, it is possible to accidentally use `TokenEndpoint.session` field in any method needed in that flow since `codeToToken` method is member method of this class. This is error-prone especially if there would be any change needed in the future. Should any method be shared across `codeToToken` and any other method in this class, it could be declared `static`.

Reply:
This sounds very well indeed.